### PR TITLE
exo: k8s cmd: deploy, delete K8S

### PR DESCRIPTION
**Pull request by pierre-emmanuelJ**
Signed-off-by: Pierre-Emmanuel Jacquier <pierre-emmanuel.jacquier@epitech.eu>

### CODE REVIEW

**Comment by pierre-emmanuelJ** [comment]
So you can deploy k8s cluster very easily !

juste type this cmd with a name and give node(s) as you want to deploy

```
Usage:
  exo k8s create <name> [flags]

Flags:
  -h, --help          help for create
  -n, --node string   node to provision [vm name | id, vm name | id,...]

Global Flags:
      --config string   Specify an alternate config file [env CLOUDSTACK_CONFIG]
  -r, --region string   config ini file section name [env CLOUDSTACK_REGION] (default "cloudstack")
```

**Comment by pierre-emmanuelJ** [comment]
@greut 
A way to deploy without specify vm but create instances depending of nodes specified can be cool !
But if you want to specify instance capacity, size... it's heavy in this cmd !

**Comment by pierre-emmanuelJ** [comment]
UPDATE:

Now you can give how much nodes you want with his capacity and BOOM !!! 🚀 

```
Create k8s cluster

Usage:
  exo k8s create <name> [flags]

Flags:
  -h, --help                   help for create
  -n, --node string            Node can provision existing instances [vm name | id, vm name | id,...]
      --node-capacity string   Node(s) capacity (if --node not set) (default "medium")
      --node-number int        Node number to create (if --node not set) (default 1)

Global Flags:
      --config string   Specify an alternate config file [env CLOUDSTACK_CONFIG]
  -r, --region string   config ini file section name [env CLOUDSTACK_REGION] (default "cloudstack")
```

File: cmd/exo/cmd/k8s_create.go [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+			if err != nil {
+				log.Fatal(err)
+			}
+
+			nodes, err = deployNodes(nodeNumber, nodeCap)
+			if err != nil {
+				log.Fatal(err)
+			}
+
+			///////MMMDDRRR J'avais pas pensé au CLOUDINIT qui job !!!!!
```
Comment (pyr) [inline cmd/exo/cmd/k8s_create.go]: leftover comment

Comment (pierre-emmanuelJ) [reply cmd/exo/cmd/k8s_create.go]: I have put a more relevant comment now

**Comment by pierre-emmanuelJ** [comment]
now just try CloudInit waiting job is ready

```
$ exo k8s create cluster-test-1 --node-number 3
Creating sshkey
Deploying node-1.....
Creating sshkey
Deploying node-2.....
Creating sshkey
Deploying node-3....
Installing Docker on node(s)................
INFO[0000] Building Kubernetes cluster
INFO[0000] [dialer] Setup tunnel for host [159.100.249.44]
INFO[0001] [dialer] Setup tunnel for host [159.100.246.20]
INFO[0002] [dialer] Setup tunnel for host [159.100.247.131]

...
```

**Comment by pierre-emmanuelJ** [comment]
UPDATE:

exo k8s delete cmd:

```
exo k8s delete cluster-test
Destroying node-1..........
70d77ebb-0eb0-49fd-a7cc-bcfc90c856a4
Destroying node-2.........
34c2f719-34e4-4f4d-b449-44ef665f9ce2
Destroying node-3..........
31844100-33af-4f03-a805-e81b502f1401
cluster-test
```

**Comment by pierre-emmanuelJ** [comment]
UPDATE:

exo k8s list:

```
$ exo k8s list
┼────────────────────┼
│ KUBERNETES CLUSTER │
┼────────────────────┼
│ cluster1           │
│ cluster2           │
┼────────────────────┼
$
```

**Comment by pierre-emmanuelJ** [comment]
UPDATE:

For this new feature you need to install or put in your path `kubectl`



exo kubectl cmd:

```
$ exo kubectl cluster-test cluster-info
Kubernetes master is running at https://159.100.250.183:6443
KubeDNS is running at https://159.100.250.183:6443/api/v1/namespaces/kube-system/services/kube-dns:dns/proxy

To further debug and diagnose cluster problems, use 'kubectl cluster-info dump'.
$
```

**Comment by pierre-emmanuelJ** [comment]
@pyr and @greut  how about to split k8s creation:

```
$ exo k8s create cluster.yml --node-number 3 --node-capacity titan
Creating sshkey
Deploying node-1...
Creating sshkey
Deploying node-2.....
Creating sshkey
Deploying node-3....

cluster.yml was created in "$PATH/cluster.yml" if you want to edit it.

run "rke up --config $PATH/cluster.yml" tu deploy kubernetes cluster

```

it can be possible to add `--auto` flag to do like before

**Comment by greut** [comment]
it's a hard problem... doing anything more than the `nodes:` section of the `cluster.yml` feels like doing too much.

https://rancher.com/docs/rancher/v2.x/en/installation/ha-server-install/

File: cmd/exo/cmd/k8s.go [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    image: rancher/k8s:v1.10.0-rancher1-2
+  kube-api:
+    image: rancher/k8s:v1.10.0-rancher1-2
+  kube-controller:
+    image: rancher/k8s:v1.10.0-rancher1-2
+  scheduler:
+    image: rancher/k8s:v1.10.0-rancher1-2
+  kubeproxy:
+    image: rancher/k8s:v1.10.0-rancher1-2
+`
```
Comment (greut) [inline cmd/exo/cmd/k8s.go]: let's stick to the minimal one for a start: https://github.com/rancher/rke#minimal-clusteryml-example

Comment (pierre-emmanuelJ) [reply cmd/exo/cmd/k8s.go]: Okay i'll watch it

Comment (greut) [reply cmd/exo/cmd/k8s.go]: I meant, remove the `services`

File: cmd/exo/cmd/k8s_create.go [hunk 3]
```diff
@@ -0,0 +1,381 @@
...
+	}
+	return nil
+}
+
+func checkingCloudInitJob(vms []string) error {
+
+	defer println("")
+
+	print("Installing Docker on node(s)")
+	for i := 0; i < 120; i++ {
```
[the remark this answers is not in the document]
Comment (pierre-emmanuelJ) [reply cmd/exo/cmd/k8s_create.go]: 👍

File: cmd/exo/cmd/k8s_create.go [hunk 4]
```diff
@@ -99,18 +99,48 @@ var k8sCreateCmd = &cobra.Command{
...
-			println(clusterFile)
+
+			msg := `
+%q file was created in:
+    %q
+
+If you want to deploy your Kubernetes cluster juste type:
+	"$ rke up --config %s"
+	
+More details in documentation here: (WIP comming soon)
```
Comment (pierre-emmanuelJ) [inline cmd/exo/cmd/k8s_create.go]: WIP

File: cmd/exo/cmd/k8s_create.go [hunk 5]
```diff
@@ -99,18 +99,48 @@ var k8sCreateCmd = &cobra.Command{
...
+If you want use kubectl program:
+    just copy %q file in "$HOME/.kube/config"
+    "$ cp %s $HOME/.kube/config"
+
+OR
+    "$ kubectl --kubeconfig %s [all kubectl commands]"
+
+If you want to connect to your web kubernetes dashboard follow this link:
+    WIP https://dfjskjskl.com 
+`
```
Comment (pierre-emmanuelJ) [inline cmd/exo/cmd/k8s_create.go]: WIP

File: cmd/exo/cmd/k8s_create.go [hunk 6]
```diff
@@ -0,0 +1,381 @@
...
+
+		zone, err := getZoneIDByName(cs, "ch-dk-2")
+		if err != nil {
+			log.Fatal(err)
+		}
+
+		template, err := getTemplateIDByName(cs, "Linux Ubuntu 18.04 LTS", zone)
+		if err != nil {
+			log.Fatal(err)
+		}
```
Comment (greut) [inline cmd/exo/cmd/k8s_create.go]: the cloud-init says xenial, I'd keep Xenial and let the user choose the region. We had this issue with Rancher before and it's been a pain.

File: cmd/exo/cmd/port.go [hunk 7]
```diff
@@ -47,8 +47,14 @@ const (
...
 	// HTTPS protocol port number
 	HTTPS port = 443
+	//ETCDCLIENT protocol port number
+	ETCDCLIENT port = 2379
+	//ETCDSERVER protocol port number
+	ETCDSERVER port = 2380
 	// RDP protocol port number
 	RDP port = 3389
+	//RKEHTTPS protocol port number
+	RKEHTTPS port = 6443
```
Comment (greut) [inline cmd/exo/cmd/port.go]: 6443 is the Kubernetes API Server, `KubernetesAPIServer`

File: cmd/exo/cmd/port.go [hunk 8]
```diff
@@ -47,8 +47,14 @@ const (
 	IRC port = 194
 	// HTTPS protocol port number
 	HTTPS port = 443
+	//ETCDCLIENT protocol port number
+	ETCDCLIENT port = 2379
+	//ETCDSERVER protocol port number
+	ETCDSERVER port = 2380
```
Comment (greut) [inline cmd/exo/cmd/port.go]: `ETCDClient`, `ETCDServer` ?

File: cmd/exo/cmd/port.go (hunk 7 above)
Comment (pierre-emmanuelJ) [reply cmd/exo/cmd/port.go]: you'r right

**Comment by greut** [comment]
https://kubeformation.sh/

**Comment by greut** [comment]
> _kubeformation_ is exclusively meant for managed Kubernetes providers.

oh.. snap!

**Comment by pierre-emmanuelJ** [comment]
oohhww but you can't choose exoscale !

**Comment by pierre-emmanuelJ** [comment]
https://kubeformation.sh/ generate volume ! why not add this

**Comment by pierre-emmanuelJ** [comment]
DOC FOR THIS CMD

# Deploy Kubernetes with EXO

## Synopsis

- In this documentation we will show you how to deploy kubernetes cluster with a simple exo command `$ exo k8s`

## Prerequisites

- Make sure you have [`rke`](https://github.com/rancher/rke/releases) and `ssh` program to your `$PATH`

## Usage

### EXO k8s command

```
$ exo k8s -h
Manage kubernetes clusters

Usage:
  exo k8s [command]

Available Commands:
  create      Create k8s cluster
  delete      Remove kubernetes cluster
  list        List kubernetes cluster(s)

Flags:
  -h, --help   help for k8s

Global Flags:
      --config string   Specify an alternate config file [env CLOUDSTACK_CONFIG]
  -r, --region string   config ini file section name [env CLOUDSTACK_REGION] (default "cloudstack")

Use "exo k8s [command] --help" for more information about a command.
$
```

- You can `create`, `delete` and  `list` your cluster(s)

### EXO k8s create command

```
$ exo k8s create -h
Create k8s cluster

Usage:
  exo k8s create <name> [flags]

Flags:
  -f, --firewall-rules-add      Add firewall rules for kubernetes (if --node not set)
  -h, --help                    help for create
      --no-auto
  -n, --node string             Node can provision existing instances [vm name | id, vm name | id,...]
      --node-capacity string    Node(s) capacity (if --node not set) (micro|tiny|small|medium|large|extra-large|huge|mega|titan) (default "medium")
      --node-number int         Node number to create (if --node not set) (default 1)
  -s, --security-group string   Create node(s) in a security group <security group name | id> (if --node not set) (default "default")

Global Flags:
      --config string   Specify an alternate config file [env CLOUDSTACK_CONFIG]
  -r, --region string   config ini file section name [env CLOUDSTACK_REGION] (default "cloudstack")
$
```

## Let's deploy !


### Configure your Exoscale firewall environement

```
$ exo firewall create k8s-security-group --description "Security group for my new kubernetes cluster"
```
### Deploy K8s

```
$ exo k8s create example-cluster --node-number 3 --security-group k8s-security-group --firewall-rules-add
```

#### this folowing command deploy k8s cluster:

- `--node-number 3` create a cluster with 3 nodes
- `--security-group k8s-security-group` put all nodes in `k8s-security-group` security group
- `--firewall-rules-add` create all firewall rule for `rke` and `kubernetes` connections in `k8s-security-group` security group

#### See the output

```
Creating sshkey
Deploying node-1...
Creating sshkey
Deploying node-2.....
Creating sshkey
Deploying node-3..
Installing Docker on node(s).....................
RKE install:
INFO[0000] Building Kubernetes cluster

......

INFO[0118] [addons] Successfully Saved addon to Kubernetes ConfigMap: rke-ingress-controller
INFO[0118] [addons] Executing deploy job..
INFO[0124] [ingress] ingress controller nginx is successfully deployed
INFO[0124] [addons] Setting up user addons
INFO[0124] [addons] no user addons defined
INFO[0124] Finished building Kubernetes cluster successfully

Your kubernetes cluster "example-cluster" is up !

"kube_config_cluster.yml" file was created to use kubectl program

If you want use kubectl program:
    just copy "kube_config_cluster.yml" file in "~/.kube/config"
    "$ cp ~/.exoscale/k8s/clusters/example-cluster/kube_config_cluster.yml ~/.kube/config"

OR
    "$ kubectl --kubeconfig ~/.exoscale/k8s/clusters/example-cluster/kube_config_cluster.yml [all kubectl commands]"

If you want to connect to your web kubernetes dashboard follow this link:
    WIP https://dfjskjskl.com
```

#### congratulation you cluster is deployed !

## Setting up kubectl

- You can folow instruction on the result output to use kubctl

```
$ cp ~/.exoscale/k8s/clusters/example-cluster/kube_config_cluster.yml ~/.kube/config

```

- Try a kubectl command

```
$ kubectl cluster-info
Kubernetes master is running at https://159.100.250.110:6443
KubeDNS is running at https://159.100.250.110:6443/api/v1/namespaces/kube-system/services/kube-dns:dns/proxy
```

## Access to kubernetes web dashboard

### Setting up the Dashboard

[Web UI (Dashboard)](https://kubernetes.io/docs/tasks/access-application-cluster/web-ui-dashboard/)

```
$ kubectl create -f https://raw.githubusercontent.com/kubernetes/dashboard/master/src/deploy/recommended/kubernetes-dashboard.yaml
```

### Admin locally

```
$ kubectl --kubeconfig ~/.kube/config proxy
Starting to server on 127.0.0.1:8001
```
or
```
$ kubectl --kubeconfig ./kube_config_cluster.yml proxy
Starting to server on 127.0.0.1:8001
```

<http://localhost:8001/api/v1/namespaces/kube-system/services/https:kubernetes-dashboard:/proxy/>

### Login using a token

To log into the dashboard, you need to authenticate as somebody or something (Service Account), a clever hack[®](https://github.com/kubernetes/dashboard/issues/2474#issuecomment-365704926) is to pick the token from the _clusterrole-aggregation-controller_.

```
$ kubectl -n kube-system describe secrets \
   `kubectl -n kube-system get secrets | awk '/clusterrole-aggregation-controller/ {print $1}'` \
       | awk '/token:/ {print $2}'
```

then copy and paste it into the token login

Congratulation you are in your kubernetes dashboard !

**Comment by pyr** [comment]
@greut we can close this one now, right?

**Comment by greut** [comment]
sure.

**Pull request closed without merging**